### PR TITLE
Fix getOptimisticStreamUnsafe() race between optimistic & locked executions

### PR DESCRIPTION
**Pull request by slfritchie**
Prior to this patch, a NullPointerException was possible with use of
getOptimisticStreamUnsafe()

1. checking getOptimisticStreamUnsafe() for null
2. using getOptimisticStreamUnsafe()'s result to call a chained method

This race condition can happen during optimistic txn execution.
Between 1 & 2, a second thread can try optimistic txn execution,
fail in a race, then *lock* the object and re-execute.  During
the re-execution, the optimistic stream pointer is set to null.

When the first thread resumes execution, step 2 will throw
NullPointerException.

I spent a while rummaging around to see if there's a deeper problem with this race, but I don't think so.  If the optimistic txn takes some action that races with the optimistic stream pointer being set to null by some other thread, that action will be undone when the optimistic txn discovers that its optimistic lock isn't valid.  Reviewers, please double-check.

### CODE REVIEW

**Comment by slfritchie** [comment]
Fixes #802

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **39502ae**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #805 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/2SFRnpiZChwhRr2ewbuSLffbftEh71ky)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **b431a66**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #805 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/znTqBy1vqV2w8909iTAwa6iAmE0wOw7t)

**Comment by slfritchie** [comment]
@dahliamalkhi I have a unit test case crafted by hand with manually-inserted scheduler yield statements.  I haven't yet crafted an automagic annotation scheme using AspectJ.

File: runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java [hunk 1]
```diff
@@ -92,14 +92,15 @@
...
-                                || o.getOptimisticStreamUnsafe()
-                                        .isStreamCurrentContextThreadCurrentContext())
-                ),
+                .access(o -> {
+                            WriteSetSMRStream stream = o.getOptimisticStreamUnsafe();
+                            return (
+                                getWriteSetEntrySize(proxy.getStreamID()) == 0 && // No updates
+                                        // And at the correct timestamp
+                                        o.getVersionUnsafe() == getSnapshotTimestamp()
+                                        && (stream == null
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java]: good catch.

**Pull request merged**